src/Microsoft.ML.TimeSeries/SrCnnEntireAnomalyDetector.cs

```diff
@@ -490,14 +495,24 @@ public void Train(double[] values, ref double[][] results)
 
                 _minimumOriginValue = Double.MaxValue;
                 _maximumOriginValue = Double.MinValue;
+                _mean = 0.0;
 
                 Array.Resize(ref _seriesToDetect, values.Length);
                 for (int i = 0; i < values.Length; ++i)
                 {
                     _seriesToDetect[i] = values[i];
                     _minimumOriginValue = Math.Min(_minimumOriginValue, values[i]);
                     _maximumOriginValue = Math.Max(_maximumOriginValue, values[i]);
+                    _mean += values[i];
+                }
+
+                _mean /= values.Length;
+                _std = 0.0;
+                for (int i = 0; i < values.Length; ++i)
+                {
+                    _std += (values[i] - _mean) * (values[i] - _mean);
                 }
+                _std = Math.Sqrt(_std / values.Length - 1);
 
```

Comment:
nit, just a general thoughts that I believe ML.Net has much more efficient way to calculate stddev, min, max and mean. If  normalize the value is needed, `context.Transforms.NormalizeMeanVariance` is much more efficient than 2 loops here.

Also, stddev can be easily calculated via one loop:

```
double sumSquare = 0;
double sum = 0;
foreach  (int i = 0; i < values.Length; ++i)
                {	        
var value = values[0]; // introduce a var to prevent duplicate so many times
                    _seriesToDetect[i] = value ;	
                    _minimumOriginValue = Math.Min(_minimumOriginValue, value );	                   
                    _maximumOriginValue = Math.Max(_maximumOriginValue, value );	                   
                    _mean += value; // note this should be sum, and it will be cleaner to have a new var called mean.
                  sumSquare += value * value
                }

var variance = (sumSquare - ((sum * sum) / length)) / (length - 1)
var stddev = Math.Sqrt(variance)
```

Reply:
nice suggestion, updated.

---
In reply to: [553427718](https://github.com/dotnet/machinelearning/pull/5579#discussion_r553427718) [](ancestors = 553427718)

Reply:
Not important, but some notes:

There's also a suitable [array normalizer](https://github.com/dotnet/machinelearning/blob/5dbfd8acac0bf798957eea122f1413209cdf07dc/src/Microsoft.ML.FastTree/Utils/VectorUtils.cs#L54-L60) in the FastTree code. But using that would entail an unwanted FastTree NuGet dependency. 

There's another [version in the GcnTransform](https://github.com/dotnet/machinelearning/blob/5dbfd8acac0bf798957eea122f1413209cdf07dc/src/Microsoft.ML.Transforms/GcnTransform.cs#L545-L563), which could work. It operates on a `ReadOnlySpan` which may let you reduce memory churn. I haven't read the anomaly detector code throughly, though I see various lists and `.ToArray()` which perhaps could be buffers and spans? Likely not worth the effort unless speed is a concern. 

Side note: For speed, most ML algos use floats instead of doubles; except at times for accumulation. You could test if using floats doesn't decrease the algo's utility, and if ok, reduce to floats.